File: calopy/src/calopy/maths/filter/SingleComponentCosinorFilter.py
import logging

import pandas as pd
import numpy as np
from CosinorPy import cosinor1

from calopy.maths.filter.Filter import Filter

SINGLE_COMPONENT_COSINOR = "Single-component cosinor"


class SingleComponentCosinorFilter(Filter):
    logger = logging.getLogger(__name__)
    type = SINGLE_COMPONENT_COSINOR

    def __init__(self, daylength: int):
        self.daylength = daylength

    def apply(self, dataFrame):
        print("single-component cosinor")
        try:
            smoothedSeries = {}
            for col in dataFrame.columns:
                series = dataFrame[col]
                datetimeIndex = series.index
                fit_results, amp, acr, statistics = cosinor1.fit_cosinor(
                    X=range(0, len(series)),
                    Y=series.reset_index(drop=True),
                    period=self.daylength,
                    plot_on=False,
                )
                series_fit = fit_results.fittedvalues
                if len(series_fit) is not len(series):
                    series_fit = self.fix_series_length(series, statistics)

                series_fit.index = datetimeIndex
                smoothedSeries[col] = series_fit
        except Exception as e:
            print("exception single-component cosinor: " + str(e))
            re = dataFrame
        frame = pd.DataFrame(smoothedSeries)
        return frame

    def get_parameter_text(self):
        txt = "daylength:" + str(self.daylength)
        return txt

    def fix_series_length(self, series_original, statistics):
        mesor = statistics["values"][0]
        amplitude = statistics["values"][1]
        phase = statistics["values"][2]
        x = np.arange(len(series_original))
        series_fitted = mesor + amplitude * np.cos((2 * np.pi * x / self.daylength) + phase)
        return pd.Series(series_fitted, index=series_original.index)

File: calojava/src/main/java/com/calopy/maths/filter/SingleComponentCosinorFilter.java
package com.calopy.maths.filter;

import java.util.ArrayList;
import java.util.List;

public class SingleComponentCosinorFilter implements CurveFittingFilter {

    public static final String TYPE = "Single-component cosinor";

    private final int daylength;

    public SingleComponentCosinorFilter(int daylength) {
        this.daylength = daylength;
    }

    @Override
    public List<Double> apply(List<Double> data) {
        int n = data.size();

        double[] cos = new double[n];
        double[] sin = new double[n];

        for (int i = 0; i < n; i++) {
            double angle = 2.0 * Math.PI * i / daylength;
            cos[i] = Math.cos(angle);
            sin[i] = Math.sin(angle);
        }

        double sumY = 0, sumCos = 0, sumSin = 0;
        double sumYCos = 0, sumYSin = 0;
        double sumCos2 = 0, sumSin2 = 0;

        for (int i = 0; i < n; i++) {
            double y = data.get(i);
            sumY += y;
            sumCos += cos[i];
            sumSin += sin[i];
            sumYCos += y * cos[i];
            sumYSin += y * sin[i];
            sumCos2 += cos[i] * cos[i];
            sumSin2 += sin[i] * sin[i];
        }

        double mesor = sumY / n;

        double betaCos = sumYCos / sumCos2;
        double betaSin = sumYSin / sumSin2;

        double amplitude = Math.sqrt(betaCos * betaCos + betaSin * betaSin);
        double phase = Math.atan2(-betaSin, betaCos);

        List<Double> fitted = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            double value = mesor +
                    amplitude * Math.cos((2.0 * Math.PI * i / daylength) + phase);
            fitted.add(value);
        }

        return fitted;
    }

    @Override
    public String getParameterText() {
        return "daylength:" + daylength;
    }
}
